os/arch/arm/src/armv7-m/up_assert.c

```diff
@@ -434,42 +437,95 @@ static void _up_assert(int errorcode)
 }
 
 #ifdef CONFIG_BINMGR_RECOVERY
+/* Recover semaphores and watchdog timer resources */
+static void recovery_task_recover(struct tcb_s *tcb)
+{
+	sem_t *sem;
+	int state;
+
+	state = tcb->task_state;
+
+	if (tcb->waitdog) {
+		(void)wd_delete(tcb->waitdog);
+		tcb->waitdog = NULL;
+	}
+
+	if (state == TSTATE_WAIT_SEM) {
+		sem = tcb->waitsem;
+		DEBUGASSERT(sem != NULL && sem->semcount < 0);
+		sem_canceled(tcb, sem);
+		sem->semcount++;
+		tcb->waitsem = NULL;
+	} else if (state == TSTATE_WAIT_MQNOTEMPTY) {
+		DEBUGASSERT(tcb->msgwaitq && tcb->msgwaitq->nwaitnotempty > 0);
+		tcb->msgwaitq->nwaitnotempty--;
+	} else if (state == TSTATE_WAIT_MQNOTFULL) {
+		DEBUGASSERT(tcb->msgwaitq && tcb->msgwaitq->nwaitnotfull > 0);
+		tcb->msgwaitq->nwaitnotfull--;
+	}
+}
+
+static void recovery_exclude_binary(struct tcb_s *tcb)
+{
+	struct tcb_s *prev;
+	struct tcb_s *next;
+
+	prev = tcb->bin_blink;
+	next = tcb->bin_flink;
+
+	while (prev) {
+		if (prev->sched_priority > FAULTMSGSENDER_PRIORITY) {
+			/* Recover semaphores and watchdog timer resources */
+			recovery_task_recover(prev);
```

Comment:
From 479 to 483 and from 491 to 495 are almost same. How about making function or macro for common?